monix-tail/shared/src/main/scala/monix/tail/Iterant.scala

```diff
@@ -1888,6 +1898,25 @@ object Iterant extends IterantInstances {
   def pure[F[_], A](a: A): Iterant[F, A] =
     now[F, A](a)
 
+  def bracket[F[_], A, B](acquire: F[A])(
+    use: A => Iterant[F, B],
+    release: (A, BracketResult) => F[Unit]
+  )(
+    implicit F: Sync[F]
+  ): Iterant[F, B] = {
+    Iterant.liftF(acquire).flatMap { a =>
+      Iterant.suspend(F.map(F.attempt(F.delay(use(a))))(_.fold(
+        ex => Iterant.raiseError[F, B](ex),
+        iterant => iterant
+      )))
+        .doOnEarlyStop(F.suspend(release(a, EarlyStop)))
+        .doOnFullConsumption {
+          case None => F.suspend(release(a, Completed))
+          case Some(ex) => F.suspend(release(a, Error(ex)))
+        }
```

Comment:
I think you're trying too much to describe `bracket` with available operations.

`bracket` is kind of a primitive operation. I would do an ugly tail-recursive loop for it because you'd then have full control over error handling and so on.

Reply:
Yeah, I kinda said that 😄 Just wanted to get tests right on the first pass